OctopusDSC/DSCResources/cOctopusEnvironment/cOctopusEnvironment.psm1

```diff
@@ -0,0 +1,238 @@
+    [string]$Url,
+    [string]$EnvironmentName,
+    [PSCredential]$OctopusCredentials = [PSCredential]::Empty,
+    [PSCredential]$OctopusApiKey = [PSCredential]::Empty
+  )
+
+  $repository = Get-OctopusClientRepository -Url $Url `
+                                            -OctopusCredentials $OctopusCredentials `
+                                            -OctopusApiKey $OctopusApiKey
+
+
+  $environment = $repository.Environments.FindByName($EnvironmentName)
+  $repository.Environments.Delete($environment)
+}
+
+function New-Environment {
+  param (
+    [string]$Url,
+    [string]$EnvironmentName,
+    [PSCredential]$OctopusCredentials = [PSCredential]::Empty,
+    [PSCredential]$OctopusApiKey = [PSCredential]::Empty
+  )
+  $repository = Get-OctopusClientRepository -Url $Url `
+                                            -OctopusCredentials $OctopusCredentials `
+                                            -OctopusApiKey $OctopusApiKey
+
+  $environment = New-Object Octopus.Client.Model.EnvironmentResource
+  $environment.Name = $EnvironmentName
+  $repository.Environments.Create($environment) | Out-Null
+}
+
+function Get-Environment {
+  param (
+    [string]$Url,
+    [string]$EnvironmentName,
+    [PSCredential]$OctopusCredentials = [PSCredential]::Empty,
+    [PSCredential]$OctopusApiKey = [PSCredential]::Empty
+  )
+
+  $repository = Get-OctopusClientRepository -Url $Url `
+                                            -OctopusCredentials $OctopusCredentials `
+                                            -OctopusApiKey $OctopusApiKey
+
+  $environment = $repository.Environments.FindByName($EnvironmentName)
+  return $environment
+}
+
+function Get-OctopusClientRepository
+{
+  param (
+    [string]$Url,
+    [string]$EnvironmentName,
+    [PSCredential]$OctopusCredentials = [PSCredential]::Empty,
+    [PSCredential]$OctopusApiKey = [PSCredential]::Empty
+  )
+
+  if ((($null -eq $OctopusCredentials) -or ($OctopusCredentials -eq [PSCredential]::Empty)) -and (($null -eq $OctopusApiKey) -or ($OctopusApiKey -eq [PSCredential]::Empty))) {
+    throw "Please provide either 'OctopusCredentials' or 'OctopusApiKey'."
+  }
+  if ((($null -ne $OctopusCredentials) -and ($OctopusCredentials -ne [PSCredential]::Empty)) -and (($null -ne $OctopusApiKey) -and ($OctopusApiKey -ne [PSCredential]::Empty))) {
+    throw "Please provide either 'OctopusCredentials' or 'OctopusApiKey', not both."
+  }
+
+  $tempFolder = [System.IO.Path]::GetTempPath()
+  $shadowCopyFolder = Join-Path $tempFolder ([Guid]::NewGuid())
+  New-Item -type Directory $shadowCopyFolder | Out-Null
+
+  $filename = "${env:ProgramFiles}\Octopus Deploy\Octopus\Newtonsoft.Json.dll"
+  $version = [System.Diagnostics.FileVersionInfo]::GetVersionInfo($filename)
+  Write-Verbose "Shadow copying '$filename' (version $($version.FileVersion)) to $shadowCopyFolder"
+  Copy-Item $filename $shadowCopyFolder
+
+  $filename = "${env:ProgramFiles}\Octopus Deploy\Octopus\Octopus.Client.dll"
```

Comment:
Pretty sure we could do this without Octopus.Client (doesn't affect whether this PR gets merged, obvs, but I'd lean towards fewest dependencies wherver we can)

Reply:
The only problem with that is we'd have to implement the OctopusClients auth model to support username/password, which is non-trivial.